File: py/dancerecital.py
# https://open.kattis.com/problems/dancerecital

# SLOW: Same algorithm as the C++ solution using permutations, but runs too
# slowly in Python.

import itertools
import sys


def dbg(str):
    print(str, file=sys.stderr)


def main():
    MaxRoutines = 10
    MaxPerformers = 26

    routines = []

    total_routines = int(input())
    for i in range(total_routines):
        routines.append(set(input()))

    routines_range = list(range(total_routines))
    routines_range_minus_one = list(range(total_routines - 1))

    distances = [[0 for c in routines_range] for r in routines_range]

    for i in routines_range:
        for j in range(i + 1, total_routines):
            intersect = routines[i].intersection(routines[j])
            diffs = len(intersect)
            distances[i][j] = diffs
            distances[j][i] = diffs

    best = sys.maxsize

    # Check all routine permutations since N is small enough.
    for p in itertools.permutations(routines_range):
        if p[0] < p[-1]:
            # Only need to check half the permutations (those that are mirror
            # images of each other will yield the same results)
            changes = 0
            for i in routines_range_minus_one:
                diffs = distances[p[i]][p[i+1]]
                changes += diffs

                # Bail early if we know this path is not optimal.
                if (changes >= best):
                    break

            if changes < best:
                best = changes

            if best == 0:
                print("0")
                sys.exit(0)

    print(str(best))


if __name__ == "__main__":
    main()

File: cpp/dancerecital.cpp
// NAME : Dance Recital
// URL  : https://open.kattis.com/problems/dancerecital
// =============================================================================
// Brute forcing all permutations is viable because N is small enough. Chose
// C++ over Python because the latter was too slow.
// =============================================================================

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int MAX_ROUTINES = 10;
const int MAX_PERFORMERS = 26;

// Figure out how many character differences exist between two strings
int diffs(string s1, string s2)
{
    int result = 0;

    for (int i = 0; i < s1.length(); i++)
    {
        for (int j = 0; j < s2.length(); j++)
        {
            if (s1[i] == s2[j])
            {
                result++;
            }
        }
    }

    return result;
}

int main()
{
    int n;
    string str;
    vector<string> routines;

    int distances[MAX_ROUTINES][MAX_ROUTINES] = {0};

    cin >> n;
    for (int i = 0; i < n; i++)
    {
        cin >> str;
        routines.push_back(str);
    }

    // Precalculate all the differenes between all string combinations. We won't
    // technically use them all (we are only searching half the permutations), but
    // this is cheap.
    for (int i = 0; i < n - 1; i++)
    {
        for (int j = i + 1; j < n; j++)
        {
            int num_diffs = diffs(routines[i], routines[j]);
            distances[i][j] = num_diffs;
            distances[j][i] = num_diffs;
        }
    }

    int ints[MAX_ROUTINES] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
    int first_idx = 0;
    int last_idx = n - 1;
    int best = 123123123;

    do
    {
        if (ints[first_idx] < ints[last_idx])
        {
            // Only consider half the permutations. If a sequence is a mirror image
            // of one we have already processed, the result will be the same and we
            // can skip it.
            int changes = 0;
            for (int i = 0; i < n - 1; i++)
            {
                changes += distances[ints[i]][ints[i+1]];
                if (changes >= best)
                {
                    break;
                }
            }

            if (changes < best)
            {
                best = changes;
            }
        }
    } while(next_permutation(ints, ints + n));

    cout << best << endl;
}